ProcessLib/TwoPhaseFlowWithPrho/TwoPhaseFlowWithPrhoLocalAssembler.h

```diff
@@ -0,0 +1,200 @@
+            _ip_data[ip]._rho_m = 0.0;
+            _ip_data[ip]._dsw_dpg = 0.0;
+            _ip_data[ip]._dsw_drho = 0.0;
+            _ip_data[ip]._drhom_dpg = 0.0;
+            _ip_data[ip]._drhom_drho = 0.0;
+            auto const& sm = _shape_matrices[ip];
+            _ip_data[ip]._integralMeasure =
+                sm.integralMeasure * sm.detJ *
+                _integration_method.getWeightedPoint(ip).getWeight();
+            _ip_data[ip]._massOperator.setZero(ShapeFunction::NPOINTS,
+                                               ShapeFunction::NPOINTS);
+            _ip_data[ip]._diffusionOperator.setZero(ShapeFunction::NPOINTS,
+                                                    ShapeFunction::NPOINTS);
+            _ip_data[ip]._massOperator.noalias() =
+                sm.N.transpose() * sm.N * _ip_data[ip]._integralMeasure;
+            _ip_data[ip]._diffusionOperator.noalias() =
+                sm.dNdx.transpose() * sm.dNdx * _ip_data[ip]._integralMeasure;
+        }
+    }
+
+    void assemble(double const t, std::vector<double> const& local_x,
+                  std::vector<double>& local_M_data,
+                  std::vector<double>& local_K_data,
+                  std::vector<double>& local_b_data) override;
+    void preTimestepConcrete(std::vector<double> const& /*local_x*/,
+                             double const /*t*/,
+                             double const /*delta_t*/) override
+    {
+        unsigned const n_integration_points =
+            _integration_method.getNumberOfPoints();
+
+        for (unsigned ip = 0; ip < n_integration_points; ip++)
+        {
+            _ip_data[ip].pushBackState();
+        }
+    }
+    Eigen::Map<const Eigen::RowVectorXd> getShapeMatrix(
+        const unsigned integration_point) const override
+    {
+        auto const& N = _shape_matrices[integration_point].N;
+
+        // assumes N is stored contiguously in memory
+        return Eigen::Map<const Eigen::RowVectorXd>(N.data(), N.size());
+    }
+
+    std::vector<double> const& getIntPtSaturation(
+        std::vector<double>& /*cache*/) const override
+    {
+        assert(_saturation.size() > 0);
+        return _saturation;
+    }
+
+    std::vector<double> const& getIntPtNonWettingPressure(
+        std::vector<double>& /*cache*/) const override
+    {
+        assert(_pressure_nonwetting.size() > 0);
+        return _pressure_nonwetting;
+    }
+
+private:
+    MeshLib::Element const& _element;
+
+    IntegrationMethod const _integration_method;
+    std::vector<ShapeMatrices> _shape_matrices;
+
+    TwoPhaseFlowWithPrhoProcessData const& _process_data;
+    std::vector<IntegrationPointData<NodalMatrixType>> _ip_data;
+    const double& molar_mass_h2o =
+        MaterialLib::PhysicalConstant::MolarMass::Water;
+    const double& molar_mass_h2 = MaterialLib::PhysicalConstant::MolarMass::H2;
+
+    std::vector<double> _saturation;
+    std::vector<double> _pressure_nonwetting;
```

Comment:
You can also store the _saturation and _pressure_nonwetting as well the _shape_matrices in the _ip_data. That would be consistent: All of the per-integration-point data is in same structure.